src/server/routers/project.ts

```diff
@@ -1,16 +1,155 @@
+      };
+    }),
+
+  edit: stageAwareProcedure
+    .input(
+      z.object({
+        params: instanceParamsSchema,
+        projectId: z.string(),
+        updatedProject: updatedProjectSchema,
+      }),
+    )
+    .mutation(
+      async ({
+        ctx,
+        input: {
+          params: { group, subGroup, instance },
+          projectId,
+          updatedProject: {
+            title,
+            description,
+            capacityUpperBound,
+            preAllocatedStudentId,
+            tags,
+            flagIds,
+          },
+        },
+      }) => {
+        if (stageCheck(ctx.stage, Stage.PROJECT_ALLOCATION)) return;
+
+        const newPreAllocatedStudentId = preAllocatedStudentId || undefined;
+
+        const prev = await ctx.db.project.findFirstOrThrow({
+          where: { id: projectId },
+          select: { preAllocatedStudentId: true },
+        });
+
+        if (!prev.preAllocatedStudentId) {
+          if (newPreAllocatedStudentId) {
+            await updateProjectAllocation(ctx.db, {
+              group,
+              subGroup,
+              instance,
+              preAllocatedStudentId: newPreAllocatedStudentId,
+              projectId: projectId,
+            });
+          }
+        } else {
+          if (!newPreAllocatedStudentId) {
+            await ctx.db.projectAllocation.delete({
+              where: {
+                allocationId: {
+                  allocationGroupId: group,
+                  allocationSubGroupId: subGroup,
+                  allocationInstanceId: instance,
+                  projectId,
+                  userId: prev.preAllocatedStudentId,
+                },
+              },
+            });
+          } else if (prev.preAllocatedStudentId !== newPreAllocatedStudentId) {
+            await ctx.db.projectAllocation.delete({
+              where: {
+                allocationId: {
+                  allocationGroupId: group,
+                  allocationSubGroupId: subGroup,
+                  allocationInstanceId: instance,
+                  projectId,
+                  userId: prev.preAllocatedStudentId,
+                },
+              },
+            });
+            await updateProjectAllocation(ctx.db, {
+              group,
+              subGroup,
+              instance,
+              preAllocatedStudentId: newPreAllocatedStudentId,
+              projectId,
+            });
+          }
+        }
+
+        await ctx.db.project.update({
+          where: { id: projectId },
+          data: {
+            title: title,
+            description: description,
+            capacityUpperBound: capacityUpperBound,
+            preAllocatedStudentId: nullable(newPreAllocatedStudentId),
+          },
+        });
+
+        await ctx.db.flagOnProject.deleteMany({
+          where: {
+            projectId,
+            AND: { flagId: { notIn: flagIds } },
+          },
+        });
+
+        await createManyFlags(ctx.db, projectId, flagIds);
```

Comment:
Rename to createProjectFlags or something similar. Not clear from name that createManyFlags creates flagOnProject rather than flags.

Reply:
Renamed to createProjectFlags